<p style='font-size: 22px; line-height: 2; margin: 10px 50px; text-align: justify; color:purple;'> <b>1. Importando pandas, numpy e plotly</b></p>
<a href = "https://pandas.pydata.org/docs/" style='font-size: 16px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'> Documentação Pandas </a>

<a href = "https://numpy.org/doc/stable/user/index.html" style='font-size: 16px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'> Documentação Numpy </a>


In [1]:
import pandas as pd
import numpy as np

<p style='font-size: 22px; line-height: 2; margin: 10px 50px; text-align: justify; color:purple;'> <b>1.1 Sobre os dados:</b>

<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'> O dataset é composto por um conjunto textos dos artigos da alura, para com isso ser possível criar um corretor ortográfico em python aplicando técnicas de processamento em linguagem natural.</p>
    

In [2]:
# Temos que criar um vocabulário para o corretor, será a base de dados da pasta dados
# com o arquivo artigos que vamos ler, passa a ler esse arquivo e armazenar
# apelidamos o arquivo de "f"
with open('./dados/artigos.txt', 'r', encoding='utf-8') as f:
    arquivos = f.read()

print(arquivos[:500])




imagem 

Temos a seguinte classe que representa um usuário no nosso sistema:

java

Para salvar um novo usuário, várias validações são feitas, como por exemplo: Ver se o nome só contém letras, [**o CPF só números**] e ver se o usuário possui no mínimo 18 anos. Veja o método que faz essa validação:

java 

Suponha agora que eu tenha outra classe, a classe `Produto`, que contém um atributo nome e eu quero fazer a mesma validação que fiz para o nome do usuário: Ver se só contém letras. E aí? Vou


<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'>Será que esse corpus é intessante para criar o corretor? será que ele tem um número de palavras suficiente?será que ele repete muitas palavras?</p>
<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'>Veremos isso agora</p>

<p style='font-size: 22px; line-height: 2; margin: 10px 50px; text-align: justify; color:purple;'> <b>2. Tokenização</b>

<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'>Para sabermos se o nosso corpus é ideal para o trabalho, deveremos saber quantas palavras possui. Afinal, um texto com dez palavras faz no máximo dez correções por exemplo. Então precisaremos ter um corpus com um grande volume de termos.</p>

In [3]:
len(arquivos)

2605046

será que esse len() está certo ?

In [4]:
texto_exemplo = 'azul e amarelo?'
len(texto_exemplo)

15

o len retorna o número de caracteres 

In [5]:
palavras_separadas = texto_exemplo.split()
palavras_separadas

['azul', 'e', 'amarelo?']

In [6]:
len(palavras_separadas)

3

O amarelo com interrogação <b> não é uma palavra </b>.

<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'> Em NLP chamamos cada um desses splits(elemento do vetor) de token, desta maneira, estamos tokenizando a frase.
    
<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'>Token é uma sequência de caracteres, separados por um limitador, que pode ser um espaço em branco, pontuação ou quebra de linhas, por exemplo.</p>

In [7]:
# renomeando adequadamente
tokens = texto_exemplo.split()
tokens

['azul', 'e', 'amarelo?']

In [8]:
len(tokens)

3

<p style='font-size: 22px; line-height: 2; margin: 10px 50px; text-align: justify; color:purple;'> <b>2.1 Refinando a tokenização</b>
    
<a href = "https://www.nltk.org/" style='font-size: 16px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'> Documentação nltk </a>
    
<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'>Um processo bem conhecido na área de NLP é o <b>nltk</b> ou Natural Language Tool Kit, que é um conjunto de ferramentas que implementa diversos métodos e algoritmos para análise textual.

<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'>Apenas relembrando nosso problema, separamos os tokens da nossa frase com split(), mas as palavras ainda estão junto com as pontuações. Então separaremos em <u>tokens de palavras e tokens de pontuação.</u>
    

In [9]:
import nltk

# word_tokenize ele separa palavras de pontuação e retorna com uma lista de tokens de palavras e pontuação
palavras_separadas = nltk.tokenize.word_tokenize(texto_exemplo)
palavras_separadas

['azul', 'e', 'amarelo', '?']

<p style='font-size: 22px; line-height: 2; margin: 10px 50px; text-align: justify; color:purple;'> <b>2.2 Separando palavras de tokens</b>
    
<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'>Precisamos descobrir apenas a quantidade de vocábulos no corpus, não queremos a pontuação.

In [10]:
# Percorrendo a string e vendo se é palavra ou pontuação
# Se for alpha numérico é string senão pontuação
'palavra'.isalpha()

True

In [11]:
'>///'.isalpha()

False

In [12]:
# A função pega a lista de tokens e retorna somente as palavras(sem pontuação)
def separa_palavras(lista_tokens):
    lista_palavras = list()    
    for token in lista_tokens:
        if token.isalpha():
            lista_palavras.append(token)
    return lista_palavras

In [13]:
# ele retirou a pontuação
separa_palavras(palavras_separadas)

['azul', 'e', 'amarelo']

<p style='font-size: 22px; line-height: 2; margin: 10px 50px; text-align: justify; color:purple;'> <b>3. Contando palavras do Corpus</b>

In [14]:
# word_tokenize ele separa palavras de pontuação e retorna com uma lista de tokens de palavras e pontuação
lista_tokens = nltk.tokenize.word_tokenize(arquivos)
lista_tokens[:20]

['imagem',
 'Temos',
 'a',
 'seguinte',
 'classe',
 'que',
 'representa',
 'um',
 'usuário',
 'no',
 'nosso',
 'sistema',
 ':',
 'java',
 'Para',
 'salvar',
 'um',
 'novo',
 'usuário',
 ',']

In [15]:
# A função pega a lista de tokens e retorna somente as palavras(sem pontuação)
def separa_palavras(lista_tokens):
    lista_palavras = list()    
    for token in lista_tokens:
        if token.isalpha():
            lista_palavras.append(token)
    return lista_palavras

In [16]:
# Como ele retirou as pontuações, sobrou apenas as palavras
# Obtendo o número de palavras
lista_palavras = separa_palavras(lista_tokens)
print('O número de palavras do corpus é de: {}'.format(len(lista_palavras)))

O número de palavras do corpus é de: 403104


<p style='font-size: 22px; line-height: 2; margin: 10px 50px; text-align: justify; color:purple;'> <b>4. Normalização</b>

In [17]:
print(lista_palavras[:5])

['imagem', 'Temos', 'a', 'seguinte', 'classe']


<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'> Precisaremos calcular quantos vocábulos únicos existem sem repetição. A palavra "Temos" possui uma letra maiúscula, que será contabilizada como uma palavra diferente; dessa maneira, o texto deve ser normalizado, transformando as palavras em minúsculas.

In [18]:
def normalizacao(lista_palavras):
    lista_normalizada = list()
    for palavra in lista_palavras:
        lista_normalizada.append(palavra.lower())
    return lista_normalizada

In [19]:
lista_normalizada = normalizacao(lista_palavras)
lista_normalizada[:5]

['imagem', 'temos', 'a', 'seguinte', 'classe']

<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'>Com a normalização, podemos ver que a palavra "Temos" agora aparece com letras  minúsculas. 

<p style='font-size: 22px; line-height: 2; margin: 10px 50px; text-align: justify; color:purple;'> <b>5. Tipos de palavras</b>
    
<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'>Agora que a lista está normalizada, precisamos retirar as palavras que são repetidas.

In [20]:
# Vamos utilizar conjuntos, o "set" mostra o conjunto da lista, sem as repetições
set([1,2,3,3,4,5,6,6,6])

{1, 2, 3, 4, 5, 6}

<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'>Esta função <u>set</u> será utlizada então para a lista normalizada para obtermos apenas as palavras sem as repetições.

In [21]:
print('O número de palavras do corpus sem repetições é de: {}'.format(len(set(lista_normalizada))))

O número de palavras do corpus sem repetições é de: 18465


<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'>Qual o número de palavras que o corretor será capaz de aprender a corrigir? <b>18465</b> palavras, apesar de temos <b>403104</b> palavras no corpus textual.
    
<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'>Segundo <a href="https://hridiomas.com/quantas-palavras-voce-precisa-saber-para-ser-fluente/#:~:text=De%203000%20a%204000%20palavras,que%20qualquer%20pessoa%20precisa%20saber.">este site</a>, é necessário saber cerca de 800 a 1.000 palavras de um idioma para estar em nível básico, 1.500 a 2.000 para o intermediário, e de 3.000 a 4.000 vocábulos para ter fluência.

<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'>Portanto, o algoritmo aprenderá a corrigir mais de 18.000 questões, é bem mais do que o suficiente para construirmos um bom corretor ortográfico, sem levar em consideração as variações entre as palavras e outros detalhes que veremos adiante.

<p style='font-size: 22px; line-height: 2; margin: 10px 50px; text-align: justify; color:purple;'> <b>6. Detalhando o corretor</b>
    
<img src='./img/logica.png'>    
<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'>Porém, o algoritmo só está resolvendo o caso da palavra “lógica”, onde apenas adicionamos uma letra entre “l” e “gica”.

<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'>Então faremos a <b>generalização</b> desta operação em nosso algoritmo, pegando a primeira parte da string chamada de lado esquerdo, adicionando uma nova letra, e concatenando com o restante da palavra no lado direito como já sabemos:
<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:blue;'><i>esquerdo (primeira parte) +** letra **+ direito (segunda parte)</i>
    
<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'>Com a operação generalizada, deveremos pensar em como geraremos a separação da palavra em duas partes para recebermos a nova letra no meio, afinal teremos situações diferentes no dia-a-dia.
<br>
<img src='./img/logica2.png'>

<p style='font-size: 22px; line-height: 2; margin: 10px 50px; text-align: justify; color:purple;'> <b>6.1 Fatiando strings</b>
<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'>Faremos uma função que é geradora dessas palavras.

In [22]:
# Releembrando fatiamento de listas no python
lista = [1,2,3]

# Do elemento de index 1 até o elemento de index 2;ele pega até o elemento anterior por isso acresc.+1
lista[1:3]

[2, 3]

In [23]:
# Releembrando fatiamento de listas no python
lista = 'lgica'

# Do elemento de index 1 até o elemento de index 2;ele pega até o elemento anterior por isso acresc.+1
lista[1:3]

'gi'

In [24]:
# Será criada uma tupla, pegando o lado esquerdo e direito da palavra

# lado direito da palavra | lado esquerdo da palavra
# Dessa maneira é possível construir um for 
(lista[:1], lista[1:])

('l', 'gica')

In [25]:
# fatias = dividir a palavra em partes (esquerda e direita por ex.)

palavra_exemplo = 'lgica'


def geradora_palavras(palavra):
    fatias = list()
    for i in range(len(palavra) + 1):
        fatias.append((lista[:i], lista[i:]))
    print(fatias)
    #palavras_geradas = insere_letras(fatias)
    
    #return palavras_geradas

In [26]:
geradora_palavras(palavra_exemplo)

[('', 'lgica'), ('l', 'gica'), ('lg', 'ica'), ('lgi', 'ca'), ('lgic', 'a'), ('lgica', '')]


<p style='font-size: 22px; line-height: 2; margin: 10px 50px; text-align: justify; color:purple;'> <b>6.2 Operação de inserção</b>

In [27]:
# E = lado esquerdo da palavras
# D = lado direito da palavras

palavra_exemplo = 'matmática'

def insere_letras(fatias):
    novas_palavras = list()
    letras = 'abcdefghihklmnopqrstuvxyzáâàãéêèẽíîìĩóôõòúûùũñç'
    for E, D in fatias:        
        for letra in letras:
            novas_palavras.append(E + letra + D)
    return novas_palavras

# Mas quem é a direita e esquerda da função ? será feita a função geradora_palavras com as fatias
def geradora_palavras(palavra):
    fatias = list()
    for i in range(len(palavra) + 1):
        fatias.append((palavra[:i], palavra[i:]))
    palavras_geradas = insere_letras(fatias) 
    return palavras_geradas

In [28]:
# Dessa maneira foi gerado um conjunto com todas as possíbilidades com as letras
palavras_geradas = geradora_palavras(palavra_exemplo)
palavras_geradas

['amatmática',
 'bmatmática',
 'cmatmática',
 'dmatmática',
 'ematmática',
 'fmatmática',
 'gmatmática',
 'hmatmática',
 'imatmática',
 'hmatmática',
 'kmatmática',
 'lmatmática',
 'mmatmática',
 'nmatmática',
 'omatmática',
 'pmatmática',
 'qmatmática',
 'rmatmática',
 'smatmática',
 'tmatmática',
 'umatmática',
 'vmatmática',
 'xmatmática',
 'ymatmática',
 'zmatmática',
 'ámatmática',
 'âmatmática',
 'àmatmática',
 'ãmatmática',
 'ématmática',
 'êmatmática',
 'èmatmática',
 'ẽmatmática',
 'ímatmática',
 'îmatmática',
 'ìmatmática',
 'ĩmatmática',
 'ómatmática',
 'ômatmática',
 'õmatmática',
 'òmatmática',
 'úmatmática',
 'ûmatmática',
 'ùmatmática',
 'ũmatmática',
 'ñmatmática',
 'çmatmática',
 'maatmática',
 'mbatmática',
 'mcatmática',
 'mdatmática',
 'meatmática',
 'mfatmática',
 'mgatmática',
 'mhatmática',
 'miatmática',
 'mhatmática',
 'mkatmática',
 'mlatmática',
 'mmatmática',
 'mnatmática',
 'moatmática',
 'mpatmática',
 'mqatmática',
 'mratmática',
 'msatmática',
 'mtatmáti

<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'>Com isso, geramos um conjunto com todas as possibilidades de correção, ou seja, a palavra separada em duas partes com inserção de letras no meio, formando novas palavras.

<p style='font-size: 22px; line-height: 2; margin: 10px 50px; text-align: justify; color:purple;'> <b>7.  Construindo a função corretor</b>

In [29]:
# A função max() do python retorna o elemento de máxima probabilidade
# a função max() vai calcular a probabilidade de cada uma das palavras geradas e retornar a maior probablidade

def corretor(palavra):
    palavras_geradas = geradora_palavras(palavra)
    palavra_correta = max(palavras_geradas, key = probabilidade)
    return palavra_correta

<p style='font-size: 22px; line-height: 2; margin: 10px 50px; text-align: justify; color:purple;'> <b>7.1  Probabilidade das palavras geradas</b>

In [30]:
# Repetindo a função acima
def corretor(palavra):
    palavras_geradas = geradora_palavras(palavra)
    palavra_correta = max(palavras_geradas, key = probabilidade)
    return palavra_correta

In [31]:
# FreqDist () = calcula a frequencia das palavras no corpus
frequencia = nltk.FreqDist(lista_normalizada)

#most_common() mostra as paalvras com maior frequencia
print(frequencia.most_common(10))
print('')

total_palavras = len(lista_normalizada)
print(total_palavras)

[('de', 15502), ('o', 14056), ('que', 12230), ('a', 11099), ('e', 10501), ('para', 7710), ('um', 6368), ('é', 5899), ('uma', 5220), ('do', 5124)]

403104


In [32]:
frequencia['lógica']

96

In [33]:
def probabilidade(palavra_gerada):
    return frequencia[palavra_gerada]/total_palavras

In [34]:
print(probabilidade('matemática'))
print(probabilidade('matematica'))

3.721124077161229e-05
0.0


In [35]:
corretor(palavra_exemplo)

'matemática'

<p style='font-size: 22px; line-height: 2; margin: 10px 50px; text-align: justify; color:purple;'> <b>8. Preparando dados de teste</b>
<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'>Foi aplicado apenas um tipo de correção do corretor, dessa maneira será medida sua acurrácia para poder comparar a acurrácia inicial com as próximas melhorias.
    
<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'>Agora, precisaremos descobrir quais são esses dados recebidos, e para isso veremos a nossa base de dados <b>palavras.txt</b> preparada para teste.

<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'>Teremos uma lista com diversas palavras em pares, onde a primeira é a correta e a segunda está errada.

<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'>O <b>avaliador()</b> precisará receber a palavra incorreta, passá-la pelo corretor e verificar se este acertou o resultado.

In [36]:
# Vai receber a base de dados palavras.txt
def cria_dados_teste(nome_arquivo):
    lista_palavras_teste = list()
    f = open(nome_arquivo, 'r', encoding = 'utf-8')   
    # separando a palavra correta da palavra errada do arquivo f
    for linha in f:       
        correta, errada = linha.split()
        lista_palavras_teste.append((correta, errada))        
    f.close()
    return lista_palavras_teste

In [37]:
lista_teste = cria_dados_teste('./dados/palavras.txt')
lista_teste[:10]

[('podemos', 'pyodemos'),
 ('esse', 'esje'),
 ('já', 'jrá'),
 ('nosso', 'nossov'),
 ('são', 'sãêo'),
 ('dos', 'dosa'),
 ('muito', 'muifo'),
 ('imagem', 'iômagem'),
 ('sua', 'ósua'),
 ('também', 'tambéùm')]

<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'>Dessa maneira temos as tuplas, sendo na esquerda a palavra correta e na direita a palavra errada.

<p style='font-size: 22px; line-height: 2; margin: 10px 50px; text-align: justify; color:purple;'> <b>9. Avaliando o corretor</b>
    
<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'>Será iterado em cada uma das linhas da lista de palavras teste, passando para a função corretor a palavra equivocada e verificar se a palavra gerada é igual a palavra correta, se for a mesma é acrescido 1 na variável acertou.

In [38]:
def avaliador(testes):
    numero_palavras = len(testes)
    acertou = 0
    for correta, errada in testes:
        palavra_corrigida = corretor(errada)
        if palavra_corrigida == correta:
            acertou += 1
    taxa_acerto = acertou/numero_palavras
    print('A taxa de acerto é de: {:.2f}% de {} palavras'.format(taxa_acerto*100, numero_palavras))

In [39]:
avaliador(lista_teste)

A taxa de acerto é de: 1.08% de 186 palavras


<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'>O corretor possui uma taxa de acerto de apenas  <b>1.08%</b> das palavras, pois ele corrige apenas um erro específicio; serão feitas outras melhorias para obter uma melhor acurrácia.

<p style='font-size: 22px; line-height: 2; margin: 10px 50px; text-align: justify; color:purple;'> <b>10. Operação de delete</b>
    
<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'>Neste passo, corrigiremos um outro tipo de erro bastante comum também além de deletar ou esquecer de digitar, que consiste em escrever uma letra a mais por engano.

<img src='./img/delete.png'>
<br>
<img src='./img/delete2.png' width=400>

    
<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:blue;'><i>ESQUERDO + DIREITO[1:]</i>
    
<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'> Pois dessa maneira ele elimina a primeira letra do lado direito, que vai concatenando com as fatias da palavra.

<p style='font-size: 22px; line-height: 2; margin: 10px 50px; text-align: justify; color:purple;'> <b>10.1 Implementando o delete de caracteres</b>

In [40]:
exemplo_palavra = 'cciência'
# ele retirará a primeira letra
print(exemplo_palavra[1:])

ciência


In [41]:
def deletando_caracteres(fatias):
    novas_palavras = list()
    for E, D in fatias:        
        novas_palavras.append(E + D[1:])
    return novas_palavras

In [42]:
# Concatenando as listas de insere_letras e deletando_caracteres
def geradora_palavras(palavra):
    fatias = list()
    for i in range(len(palavra) + 1):
        fatias.append((palavra[:i], palavra[i:]))
    palavras_geradas = insere_letras(fatias)
    palavras_geradas += deletando_caracteres(fatias)
    return palavras_geradas

In [43]:
geradora_palavras(exemplo_palavra)

['acciência',
 'bcciência',
 'ccciência',
 'dcciência',
 'ecciência',
 'fcciência',
 'gcciência',
 'hcciência',
 'icciência',
 'hcciência',
 'kcciência',
 'lcciência',
 'mcciência',
 'ncciência',
 'occiência',
 'pcciência',
 'qcciência',
 'rcciência',
 'scciência',
 'tcciência',
 'ucciência',
 'vcciência',
 'xcciência',
 'ycciência',
 'zcciência',
 'ácciência',
 'âcciência',
 'àcciência',
 'ãcciência',
 'écciência',
 'êcciência',
 'ècciência',
 'ẽcciência',
 'ícciência',
 'îcciência',
 'ìcciência',
 'ĩcciência',
 'ócciência',
 'ôcciência',
 'õcciência',
 'òcciência',
 'úcciência',
 'ûcciência',
 'ùcciência',
 'ũcciência',
 'ñcciência',
 'çcciência',
 'caciência',
 'cbciência',
 'ccciência',
 'cdciência',
 'ceciência',
 'cfciência',
 'cgciência',
 'chciência',
 'ciciência',
 'chciência',
 'ckciência',
 'clciência',
 'cmciência',
 'cnciência',
 'cociência',
 'cpciência',
 'cqciência',
 'crciência',
 'csciência',
 'ctciência',
 'cuciência',
 'cvciência',
 'cxciência',
 'cyciência',
 'czci

In [44]:
# Testando o corretor com letra pra mais e com letra faltante
print(corretor('estatístiaca'))
print(corretor('estatístca'))

estatística
estatística


<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'>Avaliando a acurrácia do modelo.

In [45]:
avaliador(lista_teste)

A taxa de acerto é de: 41.40% de 186 palavras


<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'>A taxa de acerto melhorou consideravelmente em relação a métrica inicial que era aproximadamente de 1%.

<p style='font-size: 22px; line-height: 2; margin: 10px 50px; text-align: justify; color:purple;'> <b>11. Trocando as letras</b>
    
<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'>Ainda existem outros tipos de erros, como quando trocamos uma letra por outra. Por exemplo, na hora de digitar o “ó” no teclado, sem querer apertamos a tecla “í” que está bem ao lado do caractere correto, gerando a palavra errada "lígica".
    
<img src='./img/trocando.png' width=800>
<br>
<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'>Na tabela abaixo há o exemplo de concatenação das fatias da palavra lógica com a letra <b>ó</b>.
<img src='./img/trocando2.png' width=450>
    
<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:blue;'><i>ESQUERDO + LETRA + DIREITO[1:]</i>
    
<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'> Pois dessa maneira ele elimina a primeira letra do lado direito para ela ser substituida por outra letra.

<p style='font-size: 22px; line-height: 2; margin: 10px 50px; text-align: justify; color:purple;'> <b>11.1 Implementando a troca de letras</b>

<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'> Serão colocadas todas as funções modificadoras das palavras em uma mesma célula para ficar mais fácil de visualizar.

In [46]:
def insere_letras(fatias):
    novas_palavras = list()
    letras = 'abcdefghihklmnopqrstuvxyzáâàãéêèẽíîìĩóôõòúûùũñç'
    for E, D in fatias:        
        for letra in letras:
            novas_palavras.append(E + letra + D)
    return novas_palavras

def deletando_caracteres(fatias):
    novas_palavras = list()
    for E, D in fatias:        
        novas_palavras.append(E + D[1:])
    return novas_palavras

def troca_letra(fatias):
    novas_palavras = list()
    letras = 'abcdefghihklmnopqrstuvxyzáâàãéêèẽíîìĩóôõòúûùũñç'
    for E, D in fatias:        
        for letra in letras:
            novas_palavras.append(E + letra + D[1:])
    return novas_palavras

# Concatenando as listas de insere_letras e deletando_caracteres e troca_letra
def geradora_palavras(palavra):
    fatias = list()
    for i in range(len(palavra) + 1):
        fatias.append((palavra[:i], palavra[i:]))
    palavras_geradas = insere_letras(fatias)
    palavras_geradas += deletando_caracteres(fatias)
    palavras_geradas += troca_letra(fatias)
    return palavras_geradas

In [47]:
# testado o algoritmo
palavra_exemplo = 'lígica'
palavras_geradas = geradora_palavras(palavra_exemplo)
print(palavras_geradas)

['alígica', 'blígica', 'clígica', 'dlígica', 'elígica', 'flígica', 'glígica', 'hlígica', 'ilígica', 'hlígica', 'klígica', 'llígica', 'mlígica', 'nlígica', 'olígica', 'plígica', 'qlígica', 'rlígica', 'slígica', 'tlígica', 'ulígica', 'vlígica', 'xlígica', 'ylígica', 'zlígica', 'álígica', 'âlígica', 'àlígica', 'ãlígica', 'élígica', 'êlígica', 'èlígica', 'ẽlígica', 'ílígica', 'îlígica', 'ìlígica', 'ĩlígica', 'ólígica', 'ôlígica', 'õlígica', 'òlígica', 'úlígica', 'ûlígica', 'ùlígica', 'ũlígica', 'ñlígica', 'çlígica', 'laígica', 'lbígica', 'lcígica', 'ldígica', 'leígica', 'lfígica', 'lgígica', 'lhígica', 'liígica', 'lhígica', 'lkígica', 'llígica', 'lmígica', 'lnígica', 'loígica', 'lpígica', 'lqígica', 'lrígica', 'lsígica', 'ltígica', 'luígica', 'lvígica', 'lxígica', 'lyígica', 'lzígica', 'láígica', 'lâígica', 'làígica', 'lãígica', 'léígica', 'lêígica', 'lèígica', 'lẽígica', 'líígica', 'lîígica', 'lìígica', 'lĩígica', 'lóígica', 'lôígica', 'lõígica', 'lòígica', 'lúígica', 'lûígica', 'lùígica'

<p style='font-size: 22px; line-height: 2; margin: 10px 50px; text-align: justify; color:purple;'> <b>12. Invertendo as letras</b>
    
<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'>Neste passo, corrigiremos um erro de digitação onde letras foram invertidas em suas posições, como na palavra escrita errada "lgóica".
<img src='./img/logica3.png'>
<br>
<img src='./img/trocando4.png'>
<img src='./img/trocando3.png' width=550>
<br>    
<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:blue;'><i>ESQUERDO + DIREITO[1] + DIREITO[0]+ DIREITO[2:]</i>

In [48]:
## Todas as funções criadas anteriormente + a função que inverte as letras
def insere_letras(fatias):
    novas_palavras = list()
    letras = 'abcdefghihklmnopqrstuvxyzáâàãéêèẽíîìĩóôõòúûùũñç'
    for E, D in fatias:        
        for letra in letras:
            novas_palavras.append(E + letra + D)
    return novas_palavras

def deletando_caracteres(fatias):
    novas_palavras = list()
    for E, D in fatias:        
        novas_palavras.append(E + D[1:])
    return novas_palavras

def troca_letra(fatias):
    novas_palavras = list()
    letras = 'abcdefghihklmnopqrstuvxyzáâàãéêèẽíîìĩóôõòúûùũñç'
    for E, D in fatias:        
        for letra in letras:
            novas_palavras.append(E + letra + D[1:])
    return novas_palavras

def inverte_letras(fatias):
    novas_palavras = list()
    for E, D in fatias:
        # Porque se for menor que 1 ele não realizará a operação
        if len(D) > 1:
            novas_palavras.append(E + D[1]+ D[0] + D[2:])
    return novas_palavras
    

# Concatenando as listas de insere_letras, deletando_caracteres, troca_letra e inverte letras
def geradora_palavras(palavra):
    fatias = list()
    for i in range(len(palavra) + 1):
        fatias.append((palavra[:i], palavra[i:]))
    palavras_geradas = insere_letras(fatias)
    palavras_geradas += deletando_caracteres(fatias)
    palavras_geradas += troca_letra(fatias)
    palavras_geradas += inverte_letras(fatias)
    return palavras_geradas

In [49]:
# testado o algoritmo
palavra_exemplo = 'lgóica'
palavras_geradas = geradora_palavras(palavra_exemplo)
palavras_geradas

['algóica',
 'blgóica',
 'clgóica',
 'dlgóica',
 'elgóica',
 'flgóica',
 'glgóica',
 'hlgóica',
 'ilgóica',
 'hlgóica',
 'klgóica',
 'llgóica',
 'mlgóica',
 'nlgóica',
 'olgóica',
 'plgóica',
 'qlgóica',
 'rlgóica',
 'slgóica',
 'tlgóica',
 'ulgóica',
 'vlgóica',
 'xlgóica',
 'ylgóica',
 'zlgóica',
 'álgóica',
 'âlgóica',
 'àlgóica',
 'ãlgóica',
 'élgóica',
 'êlgóica',
 'èlgóica',
 'ẽlgóica',
 'ílgóica',
 'îlgóica',
 'ìlgóica',
 'ĩlgóica',
 'ólgóica',
 'ôlgóica',
 'õlgóica',
 'òlgóica',
 'úlgóica',
 'ûlgóica',
 'ùlgóica',
 'ũlgóica',
 'ñlgóica',
 'çlgóica',
 'lagóica',
 'lbgóica',
 'lcgóica',
 'ldgóica',
 'legóica',
 'lfgóica',
 'lggóica',
 'lhgóica',
 'ligóica',
 'lhgóica',
 'lkgóica',
 'llgóica',
 'lmgóica',
 'lngóica',
 'logóica',
 'lpgóica',
 'lqgóica',
 'lrgóica',
 'lsgóica',
 'ltgóica',
 'lugóica',
 'lvgóica',
 'lxgóica',
 'lygóica',
 'lzgóica',
 'lágóica',
 'lâgóica',
 'làgóica',
 'lãgóica',
 'légóica',
 'lêgóica',
 'lègóica',
 'lẽgóica',
 'lígóica',
 'lîgóica',
 'lìgóica',
 'lĩ

In [50]:
# Avaliação do corretor

def corretor(palavra):
    palavras_geradas = geradora_palavras(palavra)
    palavra_correta = max(palavras_geradas, key = probabilidade)
    return palavra_correta

def avaliador(testes):
    numero_palavras = len(testes)
    acertou = 0
    for correta, errada in testes:
        palavra_corrigida = corretor(errada)
        if palavra_corrigida == correta:
            acertou += 1
    taxa_acerto = acertou/numero_palavras
    print('A taxa de acerto é de: {:.2f}% de {} palavras'.format(taxa_acerto*100, numero_palavras))

In [51]:
avaliador(lista_teste)

A taxa de acerto é de: 76.88% de 186 palavras


<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'>O resultado será de <b>76.88%</b> de 186 palavras. Portanto, melhoramos consideravelmente o nosso corretor, pois começamos com aproximadamente <b>1%</b> de taxa de acerto apenas, o que era inaceitável. O erro está associado ao corretor pois ele desconhece o restante das palavras, pois se o corretor não conhece a palavra que precisa corrigir, ele não fará a correção.

<p style='font-size: 22px; line-height: 2; margin: 10px 50px; text-align: justify; color:purple;'> <b>13. Palavras desconhecidas ao vocabulário</b>
    
<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'>A taxa de erros está associada a dois fatores; <b>restrição do algoritmo</b> que não gera a palavra correta e <b>dados de treinamento</b>.

<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'>Às vezes um determinado termo ainda é <b>desconhecido</b>, e portanto não podemos realizar a correção de algo que não conhecemos.

In [52]:
def avaliador(testes, vocabulario):
    numero_palavras = len(testes)
    acertou = 0
    desconhecida = 0
    for correta, errada in testes:
        palavra_corrigida = corretor(errada)
        if palavra_corrigida == correta:
            acertou += 1
        else:
            desconhecida += (correta not in vocabulario)
    taxa_acerto = acertou/numero_palavras
    taxa_desconhecida = desconhecida/numero_palavras
    print('A taxa de acerto é de: {:.2f}% de {} palavras'.format(taxa_acerto*100, numero_palavras))
    print('A taxa de palavras desconhecidas é de: {:.2f}% '.format(taxa_desconhecida*100, numero_palavras))

In [53]:
vocabulario = set(lista_normalizada)
avaliador(lista_teste, vocabulario)

A taxa de acerto é de: 76.88% de 186 palavras
A taxa de palavras desconhecidas é de: 6.99% 


<p style='font-size: 22px; line-height: 2; margin: 10px 50px; text-align: justify; color:purple;'> <b>14. Turbinando o gerador de palavras</b>
    
<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'>Por exemplo, se escrevermos erroneamente <b><i>“lóiigica”</i></b>, precisaremos deletar o caractere “i” duas vezes, e o nosso corretor ainda não é capaz de fazer isso.

<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'>Logo, notaremos que o nosso corretor só corrige palavras que estão a uma única distância da correta. Então construiremos uma função que corrigirá as que estão a duas distâncias do resultado correto, e verificará qual será o retorno do novo corretor.

In [54]:
palavra = 'lóiigica'
# A função gerador de palavras será aplicada duas vezes, para remover os dois "ii" como exemplo da palavra acima
def gerador_turbinado(palavras_geradas):
    novas_palavras = list()
    for palavra in palavras_geradas:
        # Como a gente tem uma lista da primeira vez que a função for aplicada, aqui terá uma concatenação das listas
        novas_palavras += geradora_palavras(palavra)
    return novas_palavras

palavras_g = gerador_turbinado(geradora_palavras(palavra))

In [55]:
# Vendo se o gerador ele retorna a palavra lógica corretamente
'lógica' in palavras_g

True

In [56]:
len(palavras_g)

787396

<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'>Essa única palavra gerou <b>787396</b> variações, precisamos de alguma maneira eliminar as palavras que não existem das palavras geradas.

<p style='font-size: 22px; line-height: 2; margin: 10px 50px; text-align: justify; color:purple;'> <b>15. Escolhendo os melhores candidatos</b>
    
<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'>Anteriormente, vimos um gigantesco número de palavras geradas para uma única correção, tanto pelo gerador_palavras() quanto pelo gerador_turbinado().

<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'>Por conta disso, calcular a <b>máxima probabilidade</b> de todas esses possíveis resultados corretos se tornou inviável, então precisaremos <b>reduzir</b> este número de candidatos.

In [61]:
def novo_corretor(palavra):
    palavras_geradas = geradora_palavras(palavra)
    palavras_geradas_turbinado = gerador_turbinado(palavras_geradas)
    todas_palavras = set(palavras_geradas + palavras_geradas_turbinado)
    candidatos = [palavra]
    for palavra in todas_palavras:
        if palavra in vocabulario:
            candidatos.append(palavra)
    print(len(candidatos))
    palavra_correta = max(candidatos, key = probabilidade)
    return palavra_correta

In [62]:
novo_corretor(palavra)

2


'lógica'

<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'>Conseguimos corrigir a palavra e temos apenas <b>2 candidatos</b>, o restante que era quase 1 milhão de palavras eram insignificantes.

In [63]:
def novo_corretor(palavra):
    palavras_geradas = geradora_palavras(palavra)
    palavras_geradas_turbinado = gerador_turbinado(palavras_geradas)
    todas_palavras = set(palavras_geradas + palavras_geradas_turbinado)
    candidatos = [palavra]
    for palavra in todas_palavras:
        if palavra in vocabulario:
            candidatos.append(palavra)
    palavra_correta = max(candidatos, key = probabilidade)
    return palavra_correta

In [64]:
novo_corretor(palavra)

'lógica'

<p style='font-size: 22px; line-height: 2; margin: 10px 50px; text-align: justify; color:purple;'> <b>16. Avaliando o resultado dos dois corretores</b>

In [70]:
def avaliador(testes, vocabulario):
    numero_palavras = len(testes)
    acertou = 0
    desconhecida = 0
    for correta, errada in testes:
        palavra_corrigida = novo_corretor(errada)
        if palavra_corrigida == correta:
            acertou += 1
            desconhecida += (correta not in vocabulario)
    taxa_acerto = acertou/numero_palavras
    taxa_desconhecida = desconhecida/numero_palavras
    print('A taxa de acerto é de: {:.2f}% de {} palavras'.format(taxa_acerto*100, numero_palavras))
    print('A taxa de palavras desconhecidas é de: {:.2f}% '.format(taxa_desconhecida*100, numero_palavras))

In [71]:
vocabulario = set(lista_normalizada)
avaliador(lista_teste, vocabulario)

A taxa de acerto é de: 55.38% de 186 palavras
A taxa de palavras desconhecidas é de: 0.54% 


<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'> O número de palavras desconhecidas se modificou, mas ele não deveria ser fixo? O contador está no <b>lugar errado!!</b>
<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'> O corretor teve uma <b>piora</b> de resultados.

In [72]:
def avaliador(testes, vocabulario):
    numero_palavras = len(testes)
    acertou = 0
    desconhecida = 0
    for correta, errada in testes:
        palavra_corrigida = novo_corretor(errada)
        desconhecida += (correta not in vocabulario)
        if palavra_corrigida == correta:
            acertou += 1
    taxa_acerto = acertou/numero_palavras
    taxa_desconhecida = desconhecida/numero_palavras
    print('A taxa de acerto é de: {:.2f}% de {} palavras'.format(taxa_acerto*100, numero_palavras))
    print('A taxa de palavras desconhecidas é de: {:.2f}% '.format(taxa_desconhecida*100, numero_palavras))

In [73]:
vocabulario = set(lista_normalizada)
avaliador(lista_teste, vocabulario)

A taxa de acerto é de: 55.38% de 186 palavras
A taxa de palavras desconhecidas é de: 6.99% 


<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'> Agora sim, o valor das palavras desconhecidas não se modificou.

<p style='font-size: 22px; line-height: 2; margin: 10px 50px; text-align: justify; color:purple;'> <b>16.1  Avaliando o resultado dos dois corretores cont</b>
    
<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'>Vimos que o <i>novo_corretor()</i> performou consideravelmente <b>pior</b> do que o nosso corretor anterior. Printaremos alguns resultados para analisarmos o porquê deste novo não estar indo tão bem quanto o outro.

In [74]:
def avaliador(testes, vocabulario):
    numero_palavras = len(testes)
    acertou = 0
    desconhecida = 0
    for correta, errada in testes:
        palavra_corrigida = novo_corretor(errada)
        desconhecida += (correta not in vocabulario)
        if palavra_corrigida == correta:
            acertou += 1
        else:
            # vamos printar a palavra equivocada, a palavra gerada pelo corretor antigo e corrigida pelo novo corretor
            print(errada + '-' + corretor(errada) + '-' + palavra_corrigida)
    taxa_acerto = acertou/numero_palavras
    taxa_desconhecida = desconhecida/numero_palavras
    print('A taxa de acerto é de: {:.2f}% de {} palavras'.format(taxa_acerto*100, numero_palavras))
    print('A taxa de palavras desconhecidas é de: {:.2f}% '.format(taxa_desconhecida*100, numero_palavras))

In [75]:
vocabulario = set(lista_normalizada)
avaliador(lista_teste, vocabulario)

esje-esse-se
sãêo-são-não
dosa-dos-do
eme-em-de
eàssa-essa-esse
daõs-das-da
céda-cada-da
noâ-no-o
enêão-então-não
tĩem-tem-em
nossah-nossa-nosso
teb-tem-de
atĩ-até-a
âem-em-de
foo-foi-o
serr-ser-se
entke-entre-então
van-vai-a
çeus-seus-seu
eû-e-de
temeo-tempo-temos
semre-sempre-ser
elaá-ela-ele
síó-só-se
siàe-site-se
seém-sem-em
peln-pelo-ele
aléra-alura-agora
tdia-dia-da
tuúo-tudo-tipo
jé-é-de
sãô-são-não
odos-dos-do
siua-sua-seu
elpe-ele-esse
teos-temos-os
eũsa-essa-esse
vjmos-vamos-temos
dms-dos-de
cava-cada-para
ános-nos-no
èaso-caso-as
túem-tem-em
daáos-dados-dos
nossk-nosso-nosso
tãer-ter-ser
vté-até-é
búm-bem-um
sçerá-será-ser
entró-entre-então
uai-vai-a
sâus-seus-seu
ìeu-seu-de
fual-qual-sua
elal-ela-ele
skó-só-se
secm-sem-em
aluéa-alura-além
dil-dia-de
sód-só-se
eúaa-aeúaa-essa
ró-só-de
dĩaz-adĩaz-da
correptor-corretor-correto
trtica-tática-prática
ewpoderamento-aewpoderamento-ewpoderamento
îgato-gato-fato
cakvalo-acakvalo-carvalho
canelac-acanelac-caneta
tênisy-atênisy-tênisy

<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'>Nosso corretor antigo fez a corrigiu para dos corretamente. Porém, o novo corretor corrige para do, o que está incorreto, pois novo_corretor gera ambos os resultados.

<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'>Porém, <b>do</b> aparece muito mais vezes do que a palavra <b>dos</b>. Como dependemos dessa <b>máxima probabilidade</b>, a função exibirá o termo que possui uma maior chance de estar correta, mesmo que a outra palavra também seja uma possibilidade correta.
    
<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:black;'><b>Muitas vezes nossos erros de digitação estão apenas a uma distância da palavra equivocada</b>, ou seja, é mais comum esquecermos de digitar ou trocarmos uma única letra.
    
<p style='font-size: 18px; line-height: 2; margin: 10px 50px; text-align: justify; color:blue;'>Embora teoricamente o novo_corretor() parecesse ser melhor, está performando bem pior do que o outro. Portanto, é mais interessante continuarmos com o antigo.

In [77]:
def avaliador(testes, vocabulario):
    numero_palavras = len(testes)
    acertou = 0
    desconhecida = 0
    for correta, errada in testes:
        palavra_corrigida = corretor(errada)
        desconhecida += (correta not in vocabulario)
        if palavra_corrigida == correta:
            acertou += 1
    taxa_acerto = acertou/numero_palavras
    taxa_desconhecida = desconhecida/numero_palavras
    print('A taxa de acerto é de: {:.2f}% de {} palavras'.format(taxa_acerto*100, numero_palavras))
    print('A taxa de palavras desconhecidas é de: {:.2f}% '.format(taxa_desconhecida*100, numero_palavras))

In [78]:
vocabulario = set(lista_normalizada)
avaliador(lista_teste, vocabulario)

A taxa de acerto é de: 76.88% de 186 palavras
A taxa de palavras desconhecidas é de: 6.99% 


In [81]:
palavra = 'lgica'
print(novo_corretor(palavra))
print(corretor(palavra))

fica
lógica


In [82]:
palavra = 'lóógica'
print(novo_corretor(palavra))
print(corretor(palavra))

lógica
lógica


In [84]:
# O corretor antigo não consegue acertar essa palavra, porém esse tipo de erro de duas distâncias é mais incomum
palavra = 'liiógica'
print(novo_corretor(palavra))
print(corretor(palavra))

lógica
aliiógica
